Tools/Taxonomy/OwaspASVS/OwaspASVSTaxonomyGenerator.cs

```diff
@@ -0,0 +1,157 @@
+                DefaultConfiguration = new ReportingConfiguration { Level = FailureLevel.Warning },
+                Relationships = this.GetRelationships(r.RelatedCwe, r.RelatedNistSP80063B),
+            }));
+
+            taxonomies.Add(toolComponent);
+
+            var tool = new Tool { Driver = new ToolComponent { Name = "OWASP ASVS v4.0.2" } };
+
+            ExternalPropertyFileReferences externalPropertyFileReferences = new ExternalPropertyFileReferences();
+            externalPropertyFileReferences.Taxonomies = new List<ExternalPropertyFileReference>();
+            externalPropertyFileReferences.Taxonomies.Add(new ExternalPropertyFileReference()
+            {
+                Guid = Constants.Guid.Cwe,
+                Location = new ArtifactLocation() { Uri = new Uri("https://raw.githubusercontent.com/sarif-standard/taxonomies/5a8490df0cee6a0e7a8d4a210f8cfffbe3a5d319/CWE_v4.4.sarif") }
+            });
+
+            externalPropertyFileReferences.Taxonomies.Add(new ExternalPropertyFileReference()
+            {
+                Guid = Constants.Guid.NistSP80063B,
+                Location = new ArtifactLocation() { Uri = new Uri("https://raw.githubusercontent.com/sarif-standard/taxonomies/5a8490df0cee6a0e7a8d4a210f8cfffbe3a5d319/NIST_SP800-63B_v1.sarif") }
+            });
+
+            Run run = new Run
+            {
+                Tool = tool,
+                Taxonomies = taxonomies,
+                ExternalPropertyFileReferences = externalPropertyFileReferences
+            };
+
+            return run;
+        }
+
+        private List<ReportingDescriptorRelationship> GetRelationships(string relationshipString, string relationshipStringNistSP80063B)
+        {
+            List<ReportingDescriptorRelationship> relationships = new List<ReportingDescriptorRelationship>();
+
+            if (!string.IsNullOrWhiteSpace(relationshipString))
+            {
+                var idList = relationshipString.Split(',', '.', ';').Select(p => p.Trim()).Where(p => p.ToLower() != "none").Where(p => !string.IsNullOrWhiteSpace(p)).Select(p => "CWE-" + p).ToList();
+
+                foreach (string id in idList)
+                {
+                    relationships.Add(new ReportingDescriptorRelationship
+                    {
+                        Target = new ReportingDescriptorReference
+                        {
+                            Id = id,
+                            ToolComponent = new ToolComponentReference { Name = "CWE", Guid = Constants.Guid.Cwe },
+                        },
+                        Kinds = new string[] { "relevant" },
+                    });
+                }
+            }
+
+            if (!string.IsNullOrWhiteSpace(relationshipStringNistSP80063B))
+            {
+                var idList = relationshipStringNistSP80063B.Split('/').Select(p => p.Trim()).Where(p => p.ToLower() != "none").Where(p => !string.IsNullOrWhiteSpace(p)).Select(p => p).ToList();
```

Comment:
>Split [](http://example.com/codeflow?start=60&length=5)

you can use removeemptyentries and remove the isnullorwhitespace

Reply:
as discussed the removeemptyentries does not remove the " "
But I have change to use the TrimEntries by the way.